diff --git a/drivers/fpga/zynqmp-fpga.c b/drivers/fpga/zynqmp-fpga.c
index b0c482f..4a91d64 100644
--- a/drivers/fpga/zynqmp-fpga.c
+++ b/drivers/fpga/zynqmp-fpga.c
@@ -48,7 +48,8 @@ static int zynqmp_fpga_ops_write(struct fpga_manager *mgr,
 {
 	struct zynqmp_fpga_priv *priv;
 	char *kbuf;
-	size_t dma_size;
+	size_t kbuf_size;
+	size_t algn_size = (size + (sizeof(u32)-1)) & (~(sizeof(u32)-1));
 	dma_addr_t dma_addr;
 	int ret;
 	const struct zynqmp_eemi_ops *eemi_ops = zynqmp_pm_get_eemi_ops();
@@ -59,25 +60,30 @@ static int zynqmp_fpga_ops_write(struct fpga_manager *mgr,
 	priv = mgr->priv;
 
 	if (mgr->flags & IXR_FPGA_ENCRYPTION_EN)
-		dma_size = size + ENCRYPTED_KEY_LEN;
+		kbuf_size = algn_size + ENCRYPTED_KEY_LEN;
 	else
-		dma_size = size;
+		kbuf_size = algn_size + sizeof(u32);
 
-	kbuf = dma_alloc_coherent(priv->dev, dma_size, &dma_addr, GFP_KERNEL);
+	kbuf = dma_alloc_coherent(priv->dev, kbuf_size, &dma_addr, GFP_KERNEL);
 	if (!kbuf)
 		return -ENOMEM;
 
 	memcpy(kbuf, buf, size);
 
-	if (mgr->flags & IXR_FPGA_ENCRYPTION_EN)
-		memcpy(kbuf + size, mgr->key, ENCRYPTED_KEY_LEN);
+	if (mgr->flags & IXR_FPGA_ENCRYPTION_EN) {
+		memcpy(kbuf + algn_size, mgr->key, ENCRYPTED_KEY_LEN);
+        }
+	else {
+		u32* size_ptr = (u32*)(kbuf + algn_size);
+		*size_ptr = (u32)size;
+	}
 
 	__flush_cache_user_range((unsigned long)kbuf,
-				 (unsigned long)kbuf + dma_size);
+				 (unsigned long)kbuf + kbuf_size);
 
-	ret = eemi_ops->fpga_load(dma_addr, dma_addr + size, mgr->flags);
+	ret = eemi_ops->fpga_load(dma_addr, dma_addr + algn_size, mgr->flags);
 
-	dma_free_coherent(priv->dev, dma_size, kbuf, dma_addr);
+	dma_free_coherent(priv->dev, kbuf_size, kbuf, dma_addr);
 
 	return ret;
 }
