diff --git arch/arm/include/asm/arch-zynqmp/hardware.h arch/arm/include/asm/arch-zynqmp/hardware.h
index ba20bac..f879c4d 100644
--- arch/arm/include/asm/arch-zynqmp/hardware.h
+++ arch/arm/include/asm/arch-zynqmp/hardware.h
@@ -22,6 +22,8 @@
 #define ZYNQMP_TCM_SIZE		0x40000
 
 #define ZYNQMP_CRL_APB_BASEADDR	0xFF5E0000
+#define ZYNQMP_CRL_APB_I2C0_REF_CTRL_CLKACT        	0x1000000
+#define ZYNQMP_CRL_APB_I2C1_REF_CTRL_CLKACT        	0x1000000
 #define ZYNQMP_CRL_APB_TIMESTAMP_REF_CTRL_CLKACT	0x1000000
 #define ZYNQMP_CRL_APB_BOOT_PIN_CTRL_OUT_EN_SHIFT	0
 #define ZYNQMP_CRL_APB_BOOT_PIN_CTRL_OUT_VAL_SHIFT	8
@@ -34,7 +36,9 @@
 struct crlapb_regs {
 	u32 reserved0[36];
 	u32 cpu_r5_ctrl; /* 0x90 */
-	u32 reserved1[37];
+	u32 reserved1[35];
+	u32 i2c0_ref_ctrl;      /* 0x120 */
+	u32 i2c1_ref_ctrl;      /* 0x124 */
 	u32 timestamp_ref_ctrl; /* 0x128 */
 	u32 reserved2[53];
 	u32 boot_mode; /* 0x200 */
diff --git board/xilinx/zynqmp/zynqmp.c board/xilinx/zynqmp/zynqmp.c
index fd80844..ff56e39 100644
--- board/xilinx/zynqmp/zynqmp.c
+++ board/xilinx/zynqmp/zynqmp.c
@@ -416,3 +416,19 @@ int checkboard(void)
 	puts("Board: Xilinx ZynqMP\n");
 	return 0;
 }
+
+void i2c_init_board(void)
+{
+#if defined(CONFIG_ZYNQ_I2C0)
+	{
+		u32 val = readl(&crlapb_base->i2c0_ref_ctrl);
+		writel((val | ZYNQMP_CRL_APB_I2C0_REF_CTRL_CLKACT), &crlapb_base->i2c0_ref_ctrl);
+	}
+#endif
+#if defined(CONFIG_ZYNQ_I2C1)
+	{
+		u32 val = readl(&crlapb_base->i2c1_ref_ctrl);
+		writel((val | ZYNQMP_CRL_APB_I2C1_REF_CTRL_CLKACT), &crlapb_base->i2c1_ref_ctrl);
+	}
+#endif
+}
