diff --git a/common/Kconfig b/common/Kconfig
index 57bd16d..df48212 100644
--- a/common/Kconfig
+++ b/common/Kconfig
@@ -199,6 +199,13 @@ config BOOTDELAY
 
 	  See doc/README.autoboot for details.
 
+config MENU_SHOW
+	bool "Show menu when automatically booting"
+	depends on AUTOBOOT
+	select CMD_BOOTMENU
+	help
+	  Show menu when automatically booting.
+
 config USE_BOOTARGS
 	bool "Enable boot arguments"
 	help
diff --git a/configs/avnet_ultra96_rev1_defconfig b/configs/avnet_ultra96_rev1_defconfig
index f4ac2b3..89fb3df 100644
--- a/configs/avnet_ultra96_rev1_defconfig
+++ b/configs/avnet_ultra96_rev1_defconfig
@@ -90,3 +90,4 @@ CONFIG_SPL_GZIP=y
 CONFIG_OF_LIBFDT_OVERLAY=y
 CONFIG_EFI_LOADER_BOUNCE_BUFFER=y
 CONFIG_OF_EMBED=y
+CONFIG_MENU_SHOW=y
diff --git a/include/configs/xilinx_zynqmp.h b/include/configs/xilinx_zynqmp.h
index 8e600f3..e15b0ed 100644
--- a/include/configs/xilinx_zynqmp.h
+++ b/include/configs/xilinx_zynqmp.h
@@ -122,6 +122,12 @@
 	"importbootenv=echo Importing environment from SD ...; " \
 		"env import -t ${loadbootenv_addr} $filesize\0" \
 	"sd_uEnvtxt_existence_test=test -e mmc $sdbootdev:$partid /uEnv.txt\0" \
+        "sd_uEnvtxt_load=" \
+		"if run sd_uEnvtxt_existence_test; then " \
+			"run loadbootenv; " \
+			"echo Loaded environment from ${bootenv}; " \
+			"run importbootenv; " \
+		"fi\0" \
 	"sata_root=if test $scsidevs -gt 0; then setenv bootargs $bootargs root=/dev/sda rw rootfstype=ext4; fi\0" \
 	"sataboot=run xilinxcmd && load scsi 0 80000 boot/Image && load scsi 0 $fdt_addr boot/system.dtb && booti 80000 - $fdt_addr\0" \
 	"netboot=run xilinxcmd && tftpboot 10000000 image.ub && bootm\0" \
@@ -129,11 +135,6 @@
 		  "sf read $kernel_addr $kernel_offset $kernel_size && " \
 		  "booti $kernel_addr - $fdt_addr\0" \
 	"uenvboot=" \
-		"if run sd_uEnvtxt_existence_test; then " \
-			"run loadbootenv; " \
-			"echo Loaded environment from ${bootenv}; " \
-			"run importbootenv; " \
-		"fi; " \
 		"if test -n $uenvcmd; then " \
 			"echo Running uenvcmd ...; " \
 			"run uenvcmd; " \
@@ -185,6 +186,8 @@
 	PARTS_DEFAULT
 #endif
 
+#define CONFIG_PREBOOT		"run sd_uEnvtxt_load"
+
 /* Monitor Command Prompt */
 /* Console I/O Buffer Size */
 #define CONFIG_SYS_CBSIZE		2048
